File: algorithmic-toolbox/algorithmicWarmUp/hugeFibonacci.js
const hugeFibonacci = (n,m) => {
  let fib1ModM = 0 % m;
  let fib2ModM = 1 % m;
  let count = [0,1];
  while (count.length >= 0) {
    const temp = fib1ModM;
    fib1ModM = fib2ModM;
    fib2ModM = (temp + fib2ModM) % m;
    // count++;
    count.push(fib2ModM);
    if (fib1ModM === 0 && fib2ModM === 1) {
      count.pop();
      count.pop();
      break;
    }
  }
  const newFib = n % count.length;
  return count[newFib];
}

const readline = require('readline')
process.stdin.setEncoding('utf8');

const rl = readline.createInterface({
  input: process.stdin,
  terminal: false,
});

const readLine = line => {
  const a = parseInt(line.split(' ')[0], 10);
  const b = parseInt(line.split(' ')[1], 10);
  console.log(hugeFibonacci(a,b));
  process.exit();
};

rl.on('line', readLine);

File: algorithmic-toolbox/algorithmicWarmUp/huge_fibonacci.py
# Uses python3
import sys

def huge_fibonacci(n,m):
    fib1_mod_m = 0 % m
    fib2_mod_m = 1 % m
    remainders = [0,1]

    while len(remainders) >= 0:
        fib1_mod_m, fib2_mod_m = fib2_mod_m, ((fib1_mod_m + fib2_mod_m) % m)
        remainders.append(fib2_mod_m)

        if fib1_mod_m == 0 and fib2_mod_m == 1:
            remainders.pop()
            remainders.pop()
            break

    period_len = len(remainders)
    new_fib = n % period_len
    return remainders[new_fib]


if __name__ == '__main__':
    input = sys.stdin.readline()
    a, b = map(int, input.split())
    print(huge_fibonacci(a, b))

